pkg/communication/client_protobuf_endpoint.go

```diff
@@ -0,0 +1,186 @@
+		Name: name,
+
+		transport:      transport, // the transport
+		MessageChannel: make(chan goproto.Message),
+		messageHandler: messageHandler, // the message parser
+	}
+
+	fmt.Println("[ProtobufEndpoint] Created endpoint " + endpoint.GetName())
+	// Start a Message Handling routine to wait for messages arriving on the transport point
+	go endpoint.waitForSingleServerMessage() // TODO: redo using MessageWaiting policy
+	return endpoint
+}
+
+// TODO:
+func CreateClientProtobufEndpoint3(name string, transport ITransport, messageHandler ProtobufMessage) ProtobufEndpoint {
+	endpoint := &ClientProtobufEndpoint{
+		Name:           name,
+		transport:      transport,                  // the transport
+		MessageChannel: make(chan goproto.Message), // the unbuffered channel where parsed messages are sent
+		// this channel will block until the message is received
+		// by the top layer
+		messageHandler: messageHandler, // the message parser
+	}
+
+	fmt.Println("[ProtobufEndpoint] Created endpoint " + endpoint.GetName())
+	// Start a Message Handling routine to wait for messages arriving on the transport point
+	go endpoint.waitForServerMessage() // TODO: redo using MessageWaiting policy
+	return endpoint
+}
+
+func (endpoint *ClientProtobufEndpoint) GetName() string {
+	return endpoint.Name
+}
+
+func (endpoint *ClientProtobufEndpoint) MessageReceiver() chan goproto.Message {
+	return endpoint.MessageChannel
+}
+
+func (endpoint *ClientProtobufEndpoint) GetMessageHandler() ProtobufMessage {
+	return endpoint.messageHandler
+}
+
+func (endpoint *ClientProtobufEndpoint) CloseEndpoint() {
+	endpoint.closeReceived = true
+	fmt.Println("["+endpoint.Name+"] : CLOSING, close=", endpoint.closeReceived)
+	//TODO: close the channel
+	//close(endpoint.MessageReceiver())
+}
+
+func (endpoint *ClientProtobufEndpoint) Send(messageToSend *EndpointMessage) {
+	fmt.Println("[" + endpoint.Name + "] : SENDING Protobuf message") // %s", messageToSend.ProtobufMessage)
+	// Marshal protobuf message to raw bytes
+	msgMarshalled, err := goproto.Marshal(messageToSend.ProtobufMessage) // marshal to byte array
+	if err != nil {
+		glog.Fatal("[ClientProtobufEndpoint] During Send - marshaling error: ", err)
+		return
+	}
+	// Send using the underlying transport
+	tmsg := &TransportMessage{
+		RawMsg: msgMarshalled,
+	}
+	endpoint.transport.Send(tmsg) // TODO: catch any exceptions during send
+}
+
+func (endpoint *ClientProtobufEndpoint) waitForServerMessage() {
+	glog.Info("[" + endpoint.Name + "] : ########## Waiting for server response #######")
+	fmt.Println("[" + endpoint.Name + "][waitForServerMessage] : ########## Waiting for server response #######")
+
+	// main loop for listening server message until its message receiver channel is closed.
+	for {
+		fmt.Println("["+endpoint.Name+"] : Waiting for server message ...", endpoint.closeReceived)
+		if endpoint.closeReceived {
+			fmt.Println("[" + endpoint.Name + "] : Endpoint is closed")
```

Comment:
Here the communication is always initialized from client side, do we have an implementation here to re-initialize communication, for example, after server restarts